server/src/main/java/org/elasticsearch/cluster/routing/ExpectedShardSizeEstimator.java

```diff
@@ -51,7 +51,12 @@ public static long getExpectedShardSize(
         } else if (shard.unassigned() && shard.recoverySource().getType() == RecoverySource.Type.SNAPSHOT) {
             return snapshotShardSizeInfo.getShardSize(shard, defaultValue);
         } else {
-            return clusterInfo.getShardSize(shard, defaultValue);
+            var shardSize = clusterInfo.getShardSize(shard.shardId(), shard.primary());
```

Comment:
Is there any case where we should also try to derive a primary's size from a replica? I guess not, because if a primary is lost, a replica is promoted to a primary (which I think is not re-allocated).